In [5]:
import pandas as pd

df = pd.read_csv("../data/raw/loan.csv")

df.head()

C:\Users\Aditya\AppData\Local\Temp\ipykernel_18344\2462762399.py:3: DtypeWarning: Columns (19,47,55,112,123,124,125,128,129,130,133,139,140,141) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/raw/loan.csv")


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,NaN,NaN,2500,2500,2500.0,36 months,13.56,84.92,C,C1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,30000,30000,30000.0,60 months,18.94,777.23,D,D2,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,5000,5000,5000.0,36 months,17.97,180.69,D,D1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,4000,4000,4000.0,36 months,18.94,146.51,D,D2,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,NaN,30000,30000,30000.0,60 months,16.14,731.78,C,C4,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN


In [6]:
df.shape
df.columns.tolist()
df["loan_status"].value_counts(dropna=False)

loan_status
Fully Paid                                             1041952
Current                                                 919695
Charged Off                                             261655
Late (31-120 days)                                       21897
In Grace Period                                           8952
Late (16-30 days)                                         3737
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     31
Name: count, dtype: int64

In [11]:
df.shape

(2260668, 145)

In [12]:

df["loan_status"].value_counts()

loan_status
Fully Paid                                             1041952
Current                                                 919695
Charged Off                                             261655
Late (31-120 days)                                       21897
In Grace Period                                           8952
Late (16-30 days)                                         3737
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     31
Name: count, dtype: int64

In [13]:
df.columns.tolist()

['id',
 'member_id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'desc',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d',
 'last_credit_pull_d',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'policy_code',
 'application_type',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'acc_now_delinq',
 'tot_coll_amt',
 'tot_cur_

In [14]:
for i, col in enumerate(df.columns):
    print(i, "→", col)

0 → id
1 → member_id
2 → loan_amnt
3 → funded_amnt
4 → funded_amnt_inv
5 → term
6 → int_rate
7 → installment
8 → grade
9 → sub_grade
10 → emp_title
11 → emp_length
12 → home_ownership
13 → annual_inc
14 → verification_status
15 → issue_d
16 → loan_status
17 → pymnt_plan
18 → url
19 → desc
20 → purpose
21 → title
22 → zip_code
23 → addr_state
24 → dti
25 → delinq_2yrs
26 → earliest_cr_line
27 → inq_last_6mths
28 → mths_since_last_delinq
29 → mths_since_last_record
30 → open_acc
31 → pub_rec
32 → revol_bal
33 → revol_util
34 → total_acc
35 → initial_list_status
36 → out_prncp
37 → out_prncp_inv
38 → total_pymnt
39 → total_pymnt_inv
40 → total_rec_prncp
41 → total_rec_int
42 → total_rec_late_fee
43 → recoveries
44 → collection_recovery_fee
45 → last_pymnt_d
46 → last_pymnt_amnt
47 → next_pymnt_d
48 → last_credit_pull_d
49 → collections_12_mths_ex_med
50 → mths_since_last_major_derog
51 → policy_code
52 → application_type
53 → annual_inc_joint
54 → dti_joint
55 → verification_status_joint


In [15]:
important_keywords = [
    "loan", "term", "int_rate", "installment",
    "grade", "emp", "home", "annual_inc",
    "verification", "purpose", "dti",
    "fico", "inq", "delinq", "revol",
    "open_acc", "pub_rec", "mort_acc",
    "pub_rec_bankruptcies", "application_type"
]

for col in df.columns:
    if any(keyword in col.lower() for keyword in important_keywords):
        print(col)

loan_amnt
term
int_rate
installment
grade
sub_grade
emp_title
emp_length
home_ownership
annual_inc
verification_status
loan_status
purpose
dti
delinq_2yrs
inq_last_6mths
mths_since_last_delinq
open_acc
pub_rec
revol_bal
revol_util
application_type
annual_inc_joint
dti_joint
verification_status_joint
acc_now_delinq
open_acc_6m
inq_fi
inq_last_12m
delinq_amnt
mort_acc
mths_since_recent_inq
mths_since_recent_revol_delinq
pub_rec_bankruptcies
revol_bal_joint
sec_app_inq_last_6mths
sec_app_mort_acc
sec_app_open_acc
sec_app_revol_util
deferral_term
hardship_loan_status
settlement_term


In [17]:
for i, col in enumerate(df.columns):
    print(i, "→", col)

0 → id
1 → member_id
2 → loan_amnt
3 → funded_amnt
4 → funded_amnt_inv
5 → term
6 → int_rate
7 → installment
8 → grade
9 → sub_grade
10 → emp_title
11 → emp_length
12 → home_ownership
13 → annual_inc
14 → verification_status
15 → issue_d
16 → loan_status
17 → pymnt_plan
18 → url
19 → desc
20 → purpose
21 → title
22 → zip_code
23 → addr_state
24 → dti
25 → delinq_2yrs
26 → earliest_cr_line
27 → inq_last_6mths
28 → mths_since_last_delinq
29 → mths_since_last_record
30 → open_acc
31 → pub_rec
32 → revol_bal
33 → revol_util
34 → total_acc
35 → initial_list_status
36 → out_prncp
37 → out_prncp_inv
38 → total_pymnt
39 → total_pymnt_inv
40 → total_rec_prncp
41 → total_rec_int
42 → total_rec_late_fee
43 → recoveries
44 → collection_recovery_fee
45 → last_pymnt_d
46 → last_pymnt_amnt
47 → next_pymnt_d
48 → last_credit_pull_d
49 → collections_12_mths_ex_med
50 → mths_since_last_major_derog
51 → policy_code
52 → application_type
53 → annual_inc_joint
54 → dti_joint
55 → verification_status_joint


In [18]:
print("===== COLUMNS 0–50 =====")
for i, col in enumerate(df.columns[:51]):
    print(i, "→", col)

print("\n===== COLUMNS 51–100 =====")
for i, col in enumerate(df.columns[51:101], start=51):
    print(i, "→", col)

print("\n===== COLUMNS 101–144 =====")
for i, col in enumerate(df.columns[101:], start=101):
    print(i, "→", col)

===== COLUMNS 0–50 =====
0 → id
1 → member_id
2 → loan_amnt
3 → funded_amnt
4 → funded_amnt_inv
5 → term
6 → int_rate
7 → installment
8 → grade
9 → sub_grade
10 → emp_title
11 → emp_length
12 → home_ownership
13 → annual_inc
14 → verification_status
15 → issue_d
16 → loan_status
17 → pymnt_plan
18 → url
19 → desc
20 → purpose
21 → title
22 → zip_code
23 → addr_state
24 → dti
25 → delinq_2yrs
26 → earliest_cr_line
27 → inq_last_6mths
28 → mths_since_last_delinq
29 → mths_since_last_record
30 → open_acc
31 → pub_rec
32 → revol_bal
33 → revol_util
34 → total_acc
35 → initial_list_status
36 → out_prncp
37 → out_prncp_inv
38 → total_pymnt
39 → total_pymnt_inv
40 → total_rec_prncp
41 → total_rec_int
42 → total_rec_late_fee
43 → recoveries
44 → collection_recovery_fee
45 → last_pymnt_d
46 → last_pymnt_amnt
47 → next_pymnt_d
48 → last_credit_pull_d
49 → collections_12_mths_ex_med
50 → mths_since_last_major_derog

===== COLUMNS 51–100 =====
51 → policy_code
52 → application_type
53 → annual_inc

Loan Status check

In [19]:
df["loan_status"].value_counts(dropna=False)

loan_status
Fully Paid                                             1041952
Current                                                 919695
Charged Off                                             261655
Late (31-120 days)                                       21897
In Grace Period                                           8952
Late (16-30 days)                                         3737
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     31
Name: count, dtype: int64

In [20]:
df["loan_status"].value_counts(normalize=True, dropna=False) * 100

loan_status
Fully Paid                                             46.090448
Current                                                40.682444
Charged Off                                            11.574234
Late (31-120 days)                                      0.968608
In Grace Period                                         0.395989
Late (16-30 days)                                       0.165305
Does not meet the credit policy. Status:Fully Paid      0.087939
Does not meet the credit policy. Status:Charged Off     0.033663
Default                                                 0.001371
Name: proportion, dtype: float64

Missing Values

In [21]:
df.isnull().sum().sort_values(ascending=False).head(30)

id                                            2260668
member_id                                     2260668
url                                           2260668
orig_projected_additional_accrued_interest    2252242
hardship_amount                               2250055
hardship_dpd                                  2250055
hardship_loan_status                          2250055
deferral_term                                 2250055
hardship_end_date                             2250055
hardship_status                               2250055
hardship_start_date                           2250055
hardship_reason                               2250055
hardship_type                                 2250055
hardship_payoff_balance_amount                2250055
hardship_last_payment_amount                  2250055
hardship_length                               2250055
payment_plan_start_date                       2250055
debt_settlement_flag_date                     2227612
settlement_term             

In [22]:
df.dtypes.value_counts()

float64    105
object      36
int64        4
Name: count, dtype: int64

In [23]:
df.duplicated().sum()

np.int64(0)

In [24]:
df.memory_usage(deep=True).sum() / (1024**3)

np.float64(5.423975926823914)

In [26]:
missing = df.isnull().sum().sort_values(ascending=False)
missing[missing > 0].head(30)



id                                            2260668
member_id                                     2260668
url                                           2260668
orig_projected_additional_accrued_interest    2252242
hardship_amount                               2250055
hardship_dpd                                  2250055
hardship_loan_status                          2250055
deferral_term                                 2250055
hardship_end_date                             2250055
hardship_status                               2250055
hardship_start_date                           2250055
hardship_reason                               2250055
hardship_type                                 2250055
hardship_payoff_balance_amount                2250055
hardship_last_payment_amount                  2250055
hardship_length                               2250055
payment_plan_start_date                       2250055
debt_settlement_flag_date                     2227612
settlement_term             

In [27]:
df[["loan_status", "loan_amnt", "annual_inc", "int_rate", "dti"]].head()

,loan_status,loan_amnt,annual_inc,int_rate,dti
0,Current,2500,55000.0,13.56,18.24
1,Current,30000,90000.0,18.94,26.52
2,Current,5000,59280.0,17.97,10.51
3,Current,4000,92000.0,18.94,16.74
4,Current,30000,57250.0,16.14,26.35


In [28]:
df["loan_status"].value_counts(dropna=False)

loan_status
Fully Paid                                             1041952
Current                                                 919695
Charged Off                                             261655
Late (31-120 days)                                       21897
In Grace Period                                           8952
Late (16-30 days)                                         3737
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     31
Name: count, dtype: int64

In [31]:
# Columns available at the time of loan application/origination
candidate_features = [
    "loan_amnt",
    "term",
    "int_rate",
    "installment",
    "grade",
    "sub_grade",
    "emp_length",
    "home_ownership",
    "annual_inc",
    "verification_status",
    "purpose",
    "addr_state",
    "dti",
    "delinq_2yrs",
    "inq_last_6mths",
    "open_acc",
    "pub_rec",
    "revol_bal",
    "revol_util",
    "total_acc",
    "fico_range_low",
    "fico_range_high",
    "application_type"
]

# Check which of these columns actually exist
available_features = [
    col for col in candidate_features
    if col in df.columns
]

missing_features = [
    col for col in candidate_features
    if col not in df.columns
]

print("Available features:")
print(available_features)

print("\nNot available in dataset:")
print(missing_features)

Available features:
['loan_amnt', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'purpose', 'addr_state', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'application_type']

Not available in dataset:
['fico_range_low', 'fico_range_high']


In [32]:
selected_features = [
    "loan_amnt",
    "term",
    "int_rate",
    "installment",
    "grade",
    "sub_grade",
    "emp_length",
    "home_ownership",
    "annual_inc",
    "verification_status",
    "purpose",
    "addr_state",
    "dti",
    "delinq_2yrs",
    "inq_last_6mths",
    "open_acc",
    "pub_rec",
    "revol_bal",
    "revol_util",
    "total_acc",
    "application_type"
]

df_model = df[selected_features + ["loan_status"]].copy()

print("Shape:", df_model.shape)
df_model.head()

Shape: (2260668, 22)


,loan_amnt,term,int_rate,installment,grade,sub_grade,emp_length,home_ownership,annual_inc,verification_status,...,dti,delinq_2yrs,inq_last_6mths,open_acc,pub_rec,revol_bal,revol_util,total_acc,application_type,loan_status
0,2500,36 months,13.56,84.92,C,C1,10+ years,RENT,55000.0,Not Verified,...,18.24,0.0,1.0,9.0,1.0,4341,10.3,34.0,Individual,Current
1,30000,60 months,18.94,777.23,D,D2,10+ years,MORTGAGE,90000.0,Source Verified,...,26.52,0.0,0.0,13.0,1.0,12315,24.2,44.0,Individual,Current
2,5000,36 months,17.97,180.69,D,D1,6 years,MORTGAGE,59280.0,Source Verified,...,10.51,0.0,0.0,8.0,0.0,4599,19.1,13.0,Individual,Current
3,4000,36 months,18.94,146.51,D,D2,10+ years,MORTGAGE,92000.0,Source Verified,...,16.74,0.0,0.0,10.0,0.0,5468,78.1,13.0,Individual,Current
4,30000,60 months,16.14,731.78,C,C4,10+ years,MORTGAGE,57250.0,Not Verified,...,26.35,0.0,0.0,12.0,0.0,829,3.6,26.0,Individual,Current


filter target statuses

In [33]:
target_statuses = [
    "Fully Paid",
    "Charged Off",
    "Default"
]

df_model = df_model[
    df_model["loan_status"].isin(target_statuses)
].copy()

print("Shape after filtering:", df_model.shape)

print("\nLoan Status:")
print(df_model["loan_status"].value_counts())

Shape after filtering: (1303638, 22)

Loan Status:
loan_status
Fully Paid     1041952
Charged Off     261655
Default             31
Name: count, dtype: int64


create Target variable 

In [34]:
df_model["default"] = (
    df_model["loan_status"]
    .isin(["Charged Off", "Default"])
    .astype(int)
)

In [36]:
df_model["default"].value_counts()

default
0    1041952
1     261686
Name: count, dtype: int64

In [ ]:
df_model["default"].value_counts()

# 0 → Fully Paid → No Default

# 1 → Charged Off / Default → Default

default
0    1041952
1     261686
Name: count, dtype: int64

Target verify

In [38]:
pd.crosstab(
    df_model["loan_status"],
    df_model["default"]
)

default,0,1
loan_status,,
Charged Off,0,261655
Default,0,31
Fully Paid,1041952,0


In [39]:
df_model.shape

(1303638, 23)

In [40]:
df_model["loan_status"].value_counts()

loan_status
Fully Paid     1041952
Charged Off     261655
Default             31
Name: count, dtype: int64

In [41]:
df_model["default"].value_counts(normalize=True) * 100

default
0    79.926483
1    20.073517
Name: proportion, dtype: float64

In [ ]:
#DATA CLEANING

Missing values check

In [45]:
missing_model = df_model.isnull().sum().sort_values(ascending=False)

missing_model[missing_model > 0]

emp_length        75457
revol_util          810
dti                 312
inq_last_6mths        1
dtype: int64

In [43]:
df_model.dtypes

loan_amnt                int64
term                    object
int_rate               float64
installment            float64
grade                   object
sub_grade               object
emp_length              object
home_ownership          object
annual_inc             float64
verification_status     object
purpose                 object
addr_state              object
dti                    float64
delinq_2yrs            float64
inq_last_6mths         float64
open_acc               float64
pub_rec                float64
revol_bal                int64
revol_util             float64
total_acc              float64
application_type        object
loan_status             object
default                  int64
dtype: object

In [44]:
categorical_cols = df_model.select_dtypes(include=["object"]).columns

for col in categorical_cols:
    print(f"\n{col}")
    print(df_model[col].value_counts(dropna=False).head(15))


term
term
36 months    988774
60 months    314864
Name: count, dtype: int64

grade
grade
B    380158
C    369937
A    226245
D    195288
E     91574
F     31485
G      8951
Name: count, dtype: int64

sub_grade
sub_grade
C1    82816
B4    80607
B5    79582
B3    79425
C2    76854
C3    72735
C4    72111
B2    71691
B1    68853
C5    65421
A5    62068
A4    50519
D1    49950
D2    43483
A1    41258
Name: count, dtype: int64

emp_length
emp_length
10+ years    428553
2 years      117825
< 1 year     104552
3 years      104204
1 year        85678
5 years       81623
4 years       78033
NaN           75457
6 years       60934
8 years       59127
7 years       58148
9 years       49504
Name: count, dtype: int64

home_ownership
home_ownership
MORTGAGE    645509
RENT        517821
OWN         139849
ANY            267
OTHER          144
NONE            48
Name: count, dtype: int64

verification_status
verification_status
Source Verified    503737
Verified           407689
Not Verified       3

In [48]:
df_model.to_csv(
    "../data/processed/loan_selected.csv",
    index=False
)

## Data Understanding Summary

- Original dataset: 2,260,668 rows and 145 columns.
- Selected modeling dataset contains 23 columns.
- After filtering to resolved loan outcomes, 1,309,638 records remain.
- Target variable: `default`.
- `Fully Paid` is mapped to 0 (No Default).
- `Charged Off` and `Default` are mapped to 1 (Default).
- Default class represents approximately 20.07% of the dataset.
- Four selected features contain missing values.
- Categorical features include loan term, grade, sub-grade, employment length, home ownership, verification status, purpose, state, and application type.